### PR TITLE
Hashnode endpoint API tests

### PR DESCRIPTION
**Pull request by Kishan-Dhakan**

### CODE REVIEW

File: internal/api/util/crypto/bls.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package crypto
+
```
Comment (stewartie4) [inline internal/api/util/crypto/bls.go:2]: Why do we need to pull all this crypto stuff in?
We haven't needed to thus far for BLS signing etc.

File: tests/api_tests/blobber_hashnode_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+		blobberRequest := &model.BlobberGetHashnodeRequest{
+			AllocationID:    allocation.ID,
+			URL:             getBlobberURL(usedBlobberID, allocation.Blobbers),
+			ClientId:        wallet.ClientID,
+			ClientKey:       wallet.ClientKey,
+			ClientSignature: sign,
+		}
+
+		getBlobberResponse, restyResponse, err := apiClient.V1BlobberGetHashNodeRoot(t, *blobberRequest, client.HttpOkStatus)
```
Comment (stewartie4) [inline tests/api_tests/blobber_hashnode_test.go:41]: any assertions on the response body?

File: tests/api_tests/blobber_hashnode_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/0chain/system_test/internal/api/model"
+	"github.com/0chain/system_test/internal/api/util/client"
+	"github.com/0chain/system_test/internal/api/util/crypto"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHashnodeRoot(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Get hashnode root from blobber should work", func(t *testing.T) {
```
Comment (stewartie4) [inline tests/api_tests/blobber_hashnode_test.go:15]: are there any failure test cases? missing required params not sent etc.?

**Review by stewartie4 — CHANGES_REQUESTED** [review]
few small queries

File: internal/api/util/crypto/signature_scheme.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	bip39 "github.com/tyler-smith/go-bip39"
+)
+
+const CryptoVersion = "1.0"
+
+// KeyPair private and publickey
+type KeyPair struct {
+	PublicKey  string `json:"public_key"`
+	PrivateKey string `json:"private_key"`
+}
```
Comment (YarikRevich) [inline internal/api/util/crypto/signature_scheme.go:17]: There is a KeyPair structure in "internal/api/model" which has the same signature. Please use it.

**Review by YarikRevich — CHANGES_REQUESTED** [review]

File: internal/api/util/crypto/signature_scheme.go (hunk 4 above)
Comment (Kishan-Dhakan) [reply internal/api/util/crypto/signature_scheme.go:17]: This causes more trouble than it handles. Check out this comment: https://github.com/0chain/system_test/pull/430#issuecomment-1264722713

File: internal/api/util/crypto/ed255190chain.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+// GetPrivateKeyAsByteArray - converts private key into byte array
+func (ed *ED255190chainScheme) GetPrivateKeyAsByteArray() ([]byte, error) {
+	// if len(b0.PrivateKey) == 0 {
+	// 	return nil, errors.New("get_private_key_as_byte_array", "cannot convert empty private key to byte array")
+	// }
+	// privateKeyBytes, err := hex.DecodeString(b0.PrivateKey)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// return privateKeyBytes, nil
```
[the remark this answers is not in the document]
Comment (Kishan-Dhakan) [reply internal/api/util/crypto/ed255190chain.go:177]: Resolved

File: tests/api_tests/blobber_hashnode_test.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		usedBlobberID := getFirstUsedStorageNodeID(allocationBlobbers.Blobbers, allocation.Blobbers)
+		require.NotZero(t, usedBlobberID, "Old blobber ID contains zero value")
+
+
+		sign, err := crypto.SignHash(crypto.Sha3256([]byte(allocation.ID)), "bls0chain", []model.KeyPair{wallet.Keys[0]})
+		require.Nil(t, err)
+
+		blobberRequest := &model.BlobberGetHashnodeRequest{
+			AllocationID:    allocation.ID,
+			URL:             getBlobberURL(usedBlobberID, allocation.Blobbers),
```
[the remark this answers is not in the document]
Comment (Kishan-Dhakan) [reply tests/api_tests/blobber_hashnode_test.go:35]: Resolved

File: internal/api/util/client/api.go [hunk 7]
```diff
@@ -423,6 +484,29 @@ func (c *APIClient) V1SCRestGetBlobber(scRestGetBlobberRequest model.SCRestGetBl
...
+	var hashnode *model.BlobberGetHashnodeResponse
+
+	headers := map[string]string{
+		"X-App-Client-Id":        blobberGetHashnodeRequest.ClientId,
+		"X-App-Client-Key":       blobberGetHashnodeRequest.ClientKey,
+		"X-App-Client-Signature": blobberGetHashnodeRequest.ClientSignature,
+		"allocation":             blobberGetHashnodeRequest.AllocationID,
+	}
+
+	url := blobberGetHashnodeRequest.URL + "/" + strings.Replace(GetHashNodeRoot, ":allocation", blobberGetHashnodeRequest.AllocationID, 1)
```
[the remark this answers is not in the document]
Comment (Kishan-Dhakan) [reply internal/api/util/client/api.go:497]: URLBuilder class is not very intuitive or readable. It needs refactoring. I will try to incorporate this

File: internal/api/util/client/api.go [hunk 8]
```diff
@@ -133,35 +152,75 @@ func (c *APIClient) selectHealthServiceProviders(networkEntrypoint string) error
...
+
+		if len(nodes.Nodes) == 0 {
+			break
+		}
+
+		for _, node := range nodes.Nodes {
+			networkDNSResponse.Blobbers = append(networkDNSResponse.Blobbers, node.BaseURL)
+		}
+		offset += limit
+	}
```
[the remark this answers is not in the document]
Comment (Kishan-Dhakan) [reply internal/api/util/client/api.go:212]: I am planning to rename networkDNSResponse to something like networkServiceProviders. This way everything will be inside one structure and still make sense.

File: internal/api/model/api.go [hunk 9]
```diff
@@ -324,6 +322,28 @@ type SCRestGetBlobberRequest struct {
...
+	AllocationID   string `gorm:"column:allocation_id" json:"allocation_id,omitempty"`
+	Type           string `gorm:"column:type" json:"type,omitempty"`
+	Name           string `gorm:"column:name" json:"name,omitempty"`
+	Path           string `gorm:"column:path" json:"path,omitempty"`
+	ContentHash    string `gorm:"column:content_hash" json:"content_hash,omitempty"`
+	MerkleRoot     string `gorm:"column:merkle_root" json:"merkle_root,omitempty"`
+	ActualFileHash string `gorm:"column:actual_file_hash" json:"actual_file_hash,omitempty"`
+	ChunkSize      int64  `gorm:"column:chunk_size" json:"chunk_size,omitempty"`
+	Size           int64  `gorm:"column:size" json:"size,omitempty"`
+	ActualFileSize int64  `gorm:"column:actual_file_size" json:"actual_file_size,omitempty"`
```
[the remark this answers is not in the document]
Comment (Kishan-Dhakan) [reply internal/api/model/api.go:340]: Resolved

File: internal/api/util/crypto/bls.go (hunk 1 above)
Comment (Kishan-Dhakan) [reply internal/api/util/crypto/bls.go:2]: I tried using the sign function we had in place, but I kept getting `bad signature` errors from blobber side.

File: internal/api/util/client/api.go [hunk 10]
```diff
@@ -88,21 +94,30 @@ func NewAPIClient(networkEntrypoint string) *APIClient {
...
+func (c *APIClient) getHealthyNodes(nodes []string, serviceProviderType int) ([]string, error) {
 	var result []string
 	for _, node := range nodes {
 		urlBuilder := NewURLBuilder()
 		if err := urlBuilder.MustShiftParse(node); err != nil {
 			return nil, err
 		}
-		formattedURL := urlBuilder.SetPath(ChainGetStats).String()
+
+		formattedURL := ""
```
[the remark this answers is not in the document]
Comment (Kishan-Dhakan) [reply internal/api/util/client/api.go:110]: Why declare the variable in global scope here?

File: tests/api_tests/blobber_hashnode_test.go (hunk 2 above)
Comment (Kishan-Dhakan) [reply tests/api_tests/blobber_hashnode_test.go:42]: Added. Although this from an empty allocation so not much to assert on. Once we get upload API working I'll add another case, left a todo for now.

File: tests/api_tests/blobber_hashnode_test.go (hunk 3 above)
Comment (Kishan-Dhakan) [reply tests/api_tests/blobber_hashnode_test.go]: Added

File: internal/api/util/crypto/bls.go (hunk 1 above)
Comment (YarikRevich) [reply internal/api/util/crypto/bls.go:2]: Yeah, I also tried to use such stuff from gosdk it seems, but it didn't work. It's too coupled and needs to be refactored to be reused

File: internal/api/util/client/api.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (YarikRevich) [reply internal/api/util/client/api.go:457]: It's created with a usage of "Fluent Interface" pattern, which is pretty useful, when we try to implement "builders", but I will try to make it better

File: internal/api/util/client/api.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (YarikRevich) [reply internal/api/util/client/api.go:110]: I don't mean global scope. Declaration of variables with "var" keyword is just more clean

**Review by YarikRevich — CHANGES_REQUESTED** [review]

**Review by YarikRevich — APPROVED** [review]

**Comment by Kishan-Dhakan** [comment]
@stewartie4 the crypto package is required because transactions and hashes are signed differently, transactions are signed using the privatekey (which is what the SignTransaction is doing), but hashes are signed using the bls0chain scheme. So we need those packages.

**Review by stewartie4 — APPROVED** [review]
LGTM

**Pull request merged**